### PR TITLE
Rethinking Permissions

### PR DESCRIPTION
**Pull request by suneettipirneni**
### The Problem(s)

Over these past days I've been thinking about the best way to go about handling permissions for commands. As I found out recently, there's no way to "lookup" commands *before* logging into the discord client. This presented a chicken-egg problem, where I was trying to access the chicken before it even hatched.

This lead me to perhaps do lookups by name whenever the command was invoked.

For example:

```typescript
const command: Command = {
  name: 'foo',
  description: 'bar',
  run(interaction) { ... }
  channelPermissions: ['general', 'bot-commands'],
  rolePermissions: ['Moderator', 'Admin']
};
```

This actually works!

Slight Problem, it's not very expandable, and it's static by nature. 

*"Well it does what we want right?, what else do we need"*

### Consider the following:

User A: "For role permissions I want the command to run if **one of** the roles is present not all of them."

User B: "For role permissions I want the command to run if **all of** the roles are present not just one of them."

Hmm, well we can split up the properties in our `Command` object right?

```typescript
interface Command {
  allRolePermissions?: string[],
  oneOfRolePermissions: string[],
}
```

Tada!, I overcame the problem by editing the interface itself. But we forgot about User C:
"I only want roles with that have write access to channel Foo."

Ok, then I'll add another property to the command:

```typescript
interface Command {
  rolesThatOnlyHaveWriteAccessToChannel: ...
}
```

This is getting silly, and I think you can see the issue here, while the above code solutions work, they're not flexible at all. Why? because you have to change the internal implementation, in order to extend functionality.

Also some additional concerns not covered:

- I don't like the error message dispatch gives, can I change it?
- What if I want to limit my command to certain channel categories?

### So what now?

If we learned anything from the situations above, it's that we need a way to lazily evaluate the inputs given to us. However we also learned that the property that is being inputted into, needs to be flexible enough to be expanded. 

### Using functions as arguments

For lazy evaluations we can pass a function into the argument rather than an array of strings. We would check permissions within the function. The function would return a `boolean` indicating if the permissions are satisfied or not. This means we'll end up with something like this:

```typescript
const myCommand: Command = {
  name: 'test',
  description: 'A test command',
  // Check if the user has a President and Vice President role.
  permissions: (interaction: CommandInteraction) => {
    const { member } = interaction;
    const rolesToCheck = ["President", "Vice President"];
    let allowed = true;
    rolesToCheck.forEach(roleName => allowed &&= member.roles.cache.some(role => role.name === roleName));
    
    // Use our own error message.
    if (!allowed) {
      interaction.reply('Invalid role permissions.');
    }
    
    return allowed;
  }
}
```

This works as well, and it actually *resolves* the issues we brought up above, it's flexible because we can plug in our own function implementation, and aren't bound to dispatch's internal implementation(s).

### Hold on, I have to type all of that out in *each* of my commands, that looks like a lot of work!

I agree, but keep in mind that functions can be treated as values. So if you don't wanna retype a commonly used permission checker, then create a const function, and pass that around instead. 

#### `inChannels(...channels: string[])`

A helper for checking if the command was sent in one of the given channels,

Example usage:

```typescript
const command: Command = {
  permissions: inChannels('general', 'bot-commands')
}
```

#### `allRoles(...roles: string[])`

A helper for checking if the command sender has all of the given roles,

Example usage:

```typescript
const command: Command = {
  permissions: hasRoles('President', 'Mobile Developer')
}
```

#### `oneOfRoles(...roles: string[])`

A helper for checking if the sender has *one of* the given roles.

Example usage:

```typescript
const command: Command = {
  permissions: oneOfRoles('President', 'Director')
}
```

### Awesome!, but what if I want to do multiple types of checks, not just one?

There's another built-in helper function just for this:

#### `checkAll(...handlers: PermissionHandler[])`

A helper to combine/aggregate other permission handlers.

So if I want to check for channels *and* roles I can do this:

```typescript
const command: Command = {
  permissions: checkAll(
    oneOfRoles('President', 'Vice President'),
    inChannels('general', 'bot-commands'),
  )
}
```

And it just works, you can keep adding more and more to `checkAll` and it won't matter as long as your function is of type `PermissionHandler`

I feel like this achieves the goal of high flexibility and versatility while also making the issue more approachable.

### CODE REVIEW

**Comment by rob-3** [comment]
I love this idea. I haven't looked at the code but I think this is a good solution.

**Comment by Armintron** [comment]
I agree, I love this idea too. Its very easy to understand and as modular as it gets

File: src/MessageDispatcher.ts [hunk 1]
```diff
@@ -16,28 +15,16 @@ export default class MessageDispatcher implements Dispatchable {
...
-      if (!hasRoles(command.allowedRoles, interaction.member)) {
-        const errMsg =
-      'You must have one of the following roles to run this command:\n'.concat(
-        ...command.allowedRoles.map((role) => `- <@&${role}>\n`)
-      );
-        interaction.reply({ content: errMsg, ephemeral: true });
+    if (command.permissions) {
+      const check = await command.permissions(interaction);
+      if (!check) {
+        // It's not gauranted that an interaction will be responded to,
```
Comment (rob-3) [inline src/MessageDispatcher.ts]: guaranteed

**Comment by rob-3** [comment]
Only thing that concerns me is the `throw`s. Do we catch things anywhere or will this just crash the bot?

**Comment by suneettipirneni** [comment]
> Only thing that concerns me is the `throw`s. Do we catch things anywhere or will this just crash the bot?

It'll crash the bot, but the places that throw represent a truly exceptional error.

**Review by rob-3 — APPROVED** [review]

**Pull request merged**